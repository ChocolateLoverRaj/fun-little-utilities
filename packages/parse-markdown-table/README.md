# Parse Markdown Table

## Usage

### `ArrayTable`

```javascript
const { createMarkdownArrayTable } = await import('parse-markdown-table')

const markdown = `
  | id | name        | email                    |
  |----|-------------|--------------------------|
  |  1 | John Doe    | john-doe@gmail.com       |
  |  2 | Peter Smith | petersmith22@outlook.com |
  |  3 | Julia Jones | jjones778@gmail.com      |
`

// this function can take string, string[], Iterable<string>, and AsyncIterable<string>
const table = await createMarkdownArrayTable(markdown)

console.info('headers', table.headers)
for await (const row of table.rows) {
  console.info('row', row)
}
```

**Output:**

```javascript
headers [ 'id', 'name', 'email' ]
row [ '1', 'John Doe', 'john-doe@gmail.com' ]
row [ '2', 'Peter Smith', 'petersmith22@outlook.com' ]
row [ '3', 'Julia Jones', 'jjones778@gmail.com' ]
```

### `ObjectTable`

```javascript
const { createMarkdownObjectTable } = await import('parse-markdown-table')

const markdown = `
  | id | name        | email                    |
  |----|-------------|--------------------------|
  |  1 | John Doe    | john-doe@gmail.com       |
  |  2 | Peter Smith | petersmith22@outlook.com |
  |  3 | Julia Jones | jjones778@gmail.com      |
`

// this function can take string, string[], Iterable<string>, or AsyncIterable<string>
const table = createMarkdownObjectTable(markdown)

for await (const row of table) {
  console.info('row', row)
}
```

**Output:**

```javascript
row { id: '1', name: 'John Doe', email: 'john-doe@gmail.com' }
row { id: '2', name: 'Peter Smith', email: 'petersmith22@outlook.com' }
row { id: '3', name: 'Julia Jones', email: 'jjones778@gmail.com' }
```

## License

[MIT](https://git.io/fxKXN) © [Hoàng Văn Khải](https://github.com/KSXGitHub)
