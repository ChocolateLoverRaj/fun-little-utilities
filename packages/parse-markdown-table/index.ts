import { CellSet, List } from 'table-parser-base'
export * from 'table-parser-base'

const getCells = (x: string) => x.split('|').map(x => x.trim())
const horizontalLineRegex = /^[|-]+$/

export class MarkdownCellTable extends CellSet<string, string> {
  constructor (text: string) {
    const trimmedText = text.trim()
    if (!trimmedText) throw new SyntaxError('Text is empty')
    const [ headerText, ...rowText ] = trimmedText.split('\n')
    const headers = getCells(headerText)
    const leftBound = headers[0] === ''
    const rightBound = headers[headers.length - 1] === ''

    const rows = rowText
      .map(x => x.trim())
      .filter(x => !horizontalLineRegex.test(x))
      .map(getCells)

    if (leftBound) {
      headers.shift()
      for (const row of rows) {
        if (row[0] !== '') throw new SyntaxError('Inconsistent left bound')
        row.shift()
      }
    }

    if (rightBound) {
      headers.pop()
      for (const row of rows) {
        if (row[row.length - 1] !== '') throw new SyntaxError('Inconsistent right bound')
        row.pop()
      }
    }

    super(headers, rows)
  }
}

export class MarkdownObjectTable extends List<string, string> {
  constructor (text: string) {
    super(new MarkdownCellTable(text))
  }
}
