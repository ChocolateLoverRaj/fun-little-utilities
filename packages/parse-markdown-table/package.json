{
  "name": "parse-markdown-table",
  "version": "0.0.2",
  "description": "Convert a markdown table from text to JavaScript objects",
  "author": "Hoàng Văn Khải <hvksmr1996@gmail.com>",
  "license": "MIT",
  "homepage": "https://ksxnodeapps.github.io/fun-little-utilities/",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/ksxnodeapps/fun-little-utilities.git"
  },
  "bugs": {
    "url": "https://github.com/ksxnodeapps/fun-little-utilities/issues"
  },
  "keywords": [
    "table",
    "markdown"
  ],
  "sideEffects": false,
  "browser": "index.mjs",
  "module": "index.mjs",
  "main": "index.js",
  "dependencies": {
    "tslib": "^1.11.1",
    "string-stream-utils": "^0.0.2",
    "table-parser-base": "^0.0.2",
    "@types/node": "^13.11.0"
  }
}
