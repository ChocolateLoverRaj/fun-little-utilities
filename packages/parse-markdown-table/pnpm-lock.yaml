dependencies:
  '@types/node': 13.5.2
  string-stream-utils: 'link:../string-stream-utils'
  table-parser-base: 'link:../table-parser-base'
  tslib: 1.10.0
lockfileVersion: 5.1
packages:
  /@types/node/13.5.2:
    dev: false
    resolution:
      integrity: sha512-Fr6a47c84PRLfd7M7u3/hEknyUdQrrBA6VoPmkze0tcflhU5UnpWEX2kn12ktA/lb+MNHSqFlSiPHIHsaErTPA==
  /tslib/1.10.0:
    dev: false
    resolution:
      integrity: sha512-qOebF53frne81cf0S9B41ByenJ3/IuH8yJKngAX35CmiZySA0khhkovshKK+jGCaMnVomla7gVlIcc3EvKPbTQ==
specifiers:
  '@types/node': ^13.5.2
  string-stream-utils: ^0.0.0
  table-parser-base: ^0.0.0
  tslib: ^1.10.0
