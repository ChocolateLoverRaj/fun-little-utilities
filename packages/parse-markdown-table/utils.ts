export type MaybeAsyncIterable<Value> = Iterable<Value> | AsyncIterable<Value>
export type Stream = MaybeAsyncIterable<string>

export async function * splitCharacterIterable (iterable: Stream, sepChar: string) {
  let acc = ''

  for await (const char of iterable) {
    if (char === sepChar) {
      yield acc
      acc = ''
    } else {
      acc += char
    }
  }

  yield acc
}

export const iterateLines = (stream: Stream) => splitCharacterIterable(stream, '\n')

export async function * trimmedChunks (stream: Stream) {
  for await (const chunk of stream) {
    const trimmed = chunk.trim()
    if (trimmed) yield trimmed
  }
}
