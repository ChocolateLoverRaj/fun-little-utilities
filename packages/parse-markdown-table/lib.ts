import { CellSet, List } from 'table-parser-base'
import { Stream, iterateLines, trimmedChunks } from './utils'

export * from 'table-parser-base'
export { Stream }

export interface TableIterationOptions {
  readonly leftBound: boolean
  readonly rightBound: boolean
}

const horizontalLineRegex = /^[|-]+$/

export async function * iterateRows (lines: Stream, options: TableIterationOptions) {
  const { leftBound, rightBound } = options

  for await (const line of lines) {
    const trimmedLine = line.trim()
    if (horizontalLineRegex.test(trimmedLine)) continue

    const row = trimmedLine.split('|')

    if (leftBound && row.shift() !== '') {
      throw new SyntaxError('Inconsistent left bound')
    }

    if (rightBound && row.pop() !== '') {
      throw new SyntaxError('Inconsistent right bound')
    }

    yield row
  }
}

export class MarkdownCellTable extends CellSet<string, string> {
  constructor (
    public readonly headers: readonly string[],
    public readonly rows: AsyncIterableIterator<readonly string[]>
  ) {
    super()
  }
}

export async function createMarkdownCellTable (stream: Stream) {
  const lineIterator = trimmedChunks(iterateLines(stream))
  const firstLineResult = await lineIterator.next()

  if (firstLineResult.done) {
    throw new SyntaxError('Text is empty')
  }

  const headers = firstLineResult.value.split('|')
    .map(x => x.trim())
    .filter(Boolean)

  const leftBound = headers[0] === ''
  const rightBound = headers[headers.length - 1] === ''
  if (leftBound) headers.shift()
  if (rightBound) headers.pop()

  const rows = iterateRows(lineIterator, { leftBound, rightBound })

  return new MarkdownCellTable(headers, rows)
}

export class MarkdownObjectTable<Title extends string>
extends List<Title, string> {}

export async function createMarkdownObjectTable (stream: Stream) {
  return new MarkdownObjectTable(await createMarkdownCellTable(stream))
}
