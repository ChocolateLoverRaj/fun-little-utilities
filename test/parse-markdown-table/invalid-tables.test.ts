import { err, tryExec } from '@tsfun/result'
import { MarkdownObjectTable } from 'parse-markdown-table'

const setup = (text: string) => tryExec(() => new MarkdownObjectTable(text))

describe('empty string', () => {
  const text = ''

  it('throws a SyntaxError', () => {
    expect(setup(text)).toEqual(err(expect.any(SyntaxError)))
  })

  it('matches snapshot', () => {
    expect(setup(text)).toMatchSnapshot()
  })
})

describe('string with naught but whitespaces', () => {
  const text = '  \n  \t '

  it('throws a SyntaxError', () => {
    expect(setup(text)).toEqual(err(expect.any(SyntaxError)))
  })

  it('matches snapshot', () => {
    expect(setup(text)).toMatchSnapshot()
  })
})

describe('inconsistent right bound', () => {
  const text = `
    | id | name        | email                    |
    |----|-------------|--------------------------|
    |  1 | John Doe    | john-doe@gmail.com
    |  2 | Peter Smith | petersmith22@outlook.com
    |  3 | Julia Jones | jjones778@gmail.com
  `

  it('throws a SyntaxError', () => {
    expect(setup(text)).toEqual(err(expect.any(SyntaxError)))
  })

  it('matches snapshot', () => {
    expect(setup(text)).toMatchSnapshot()
  })
})

describe('inconsistent left bound', () => {
  const text = `
    | id | name        | email                    |
    |----|-------------|--------------------------|
       1 | John Doe    | john-doe@gmail.com       |
       2 | Peter Smith | petersmith22@outlook.com |
    |  3 | Julia Jones | jjones778@gmail.com      |
  `

  it('throws a SyntaxError', () => {
    expect(setup(text)).toEqual(err(expect.any(SyntaxError)))
  })

  it('matches snapshot', () => {
    expect(setup(text)).toMatchSnapshot()
  })
})
