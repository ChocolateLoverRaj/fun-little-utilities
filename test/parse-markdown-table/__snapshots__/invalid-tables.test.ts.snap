// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`empty string matches snapshot 1`] = `[SyntaxError: Text is empty]`;

exports[`inconsistent left bound matches snapshot 1`] = `[SyntaxError: Inconsistent left bound]`;

exports[`inconsistent right bound matches snapshot 1`] = `[SyntaxError: Inconsistent right bound]`;

exports[`string with naught but whitespaces matches snapshot 1`] = `[SyntaxError: Text is empty]`;
